ports/stm/supervisor/port.c

```diff
@@ -34,22 +34,151 @@
+extern uint32_t _ld_d1_ram_data_size;
+extern uint32_t _ld_d1_ram_data_flash_copy;
+extern uint32_t _ld_dtcm_bss_start;
+extern uint32_t _ld_dtcm_bss_size;
+extern uint32_t _ld_dtcm_data_destination;
+extern uint32_t _ld_dtcm_data_size;
+extern uint32_t _ld_dtcm_data_flash_copy;
+extern uint32_t _ld_itcm_destination;
+extern uint32_t _ld_itcm_size;
+extern uint32_t _ld_itcm_flash_copy;
+
+extern void main(void);
+extern void SystemInit(void);
+
+// This replaces the Reset_Handler in startup_*.S and SystemInit in system_*.c.
+__attribute__((used, naked)) void Reset_Handler(void) {
+    __disable_irq();
+    __set_MSP((uint32_t) &_ld_stack_top);
+
+    // TODO: Is any of this commented stuff actually required? 
+
+    /* Disable I cache and D cache */ 
+    // SCB_DisableICache();
+    // SCB_DisableDCache(); // this causes an instant hardfault if used
+
+    // #if ((__FPU_PRESENT == 1) && (__FPU_USED == 1))
+    //   SCB->CPACR |= ((3UL << 10*2) | (3UL << 11*2));    /* set CP10, CP11 Full Access */
+    // #endif /* ((__FPU_PRESENT == 1) && (__FPU_USED == 1)) */
+
+    // /* Disable Systick which might be enabled by bootrom */
+    // if (SysTick->CTRL & SysTick_CTRL_ENABLE_Msk)
+    // {
+    //     SysTick->CTRL &= ~SysTick_CTRL_ENABLE_Msk;
+    // }
+
+    /* Disable MPU */
+    ARM_MPU_Disable();
+
+    // Copy all of the itcm code to run from ITCM. Do this while the MPU is disabled because we write
+    // protect it.
+    for (uint32_t i = 0; i < ((size_t) &_ld_itcm_size) / 4; i++) {
+        (&_ld_itcm_destination)[i] = (&_ld_itcm_flash_copy)[i];
+    }
+
+    // The first number in RBAR is the region number. When searching for a policy, the region with
+    // the highest number wins. If none match, then the default policy set at enable applies.
+    // TODO: what is the default policy? Where is that set? 
+
+    // TODO: do I need to subdivide this up? 
+    // Mark all the flash the same until instructed otherwise. 
+    MPU->RBAR = ARM_MPU_RBAR(11, 0x08000000U);
+    MPU->RASR = ARM_MPU_RASR(EXECUTION, ARM_MPU_AP_FULL, NORMAL, NOT_SHAREABLE, CACHEABLE, BUFFERABLE, NO_SUBREGIONS, ARM_MPU_REGION_SIZE_2MB);
+
+    // This the ITCM. Set it to read-only because we've loaded everything already and it's easy to
+    // accidentally write the wrong value to 0x00000000 (aka NULL).
+    MPU->RBAR = ARM_MPU_RBAR(12, 0x00000000U);
+    MPU->RASR = ARM_MPU_RASR(EXECUTION, ARM_MPU_AP_RO, NORMAL, NOT_SHAREABLE, CACHEABLE, BUFFERABLE, NO_SUBREGIONS, ARM_MPU_REGION_SIZE_128KB);
+
+    // This the DTCM.
+    MPU->RBAR = ARM_MPU_RBAR(14, 0x20000000U);
+    MPU->RASR = ARM_MPU_RASR(EXECUTION, ARM_MPU_AP_FULL, NORMAL, NOT_SHAREABLE, CACHEABLE, BUFFERABLE, NO_SUBREGIONS, ARM_MPU_REGION_SIZE_128KB);
+
+    // This is AXI SRAM (D1).
+    MPU->RBAR = ARM_MPU_RBAR(15, 0x24000000U);
+    MPU->RASR = ARM_MPU_RASR(EXECUTION, ARM_MPU_AP_FULL, NORMAL, NOT_SHAREABLE, CACHEABLE, BUFFERABLE, NO_SUBREGIONS, ARM_MPU_REGION_SIZE_512KB);
```

Comment:
This is 384k on the h743

Reply:
It's 512k on the H743, 384 on the H742, 24/357 DS12110 Rev 7